src/com/xilinx/rapidwright/design/tools/RelocationTools.java

```diff
@@ -0,0 +1,101 @@
+package com.xilinx.rapidwright.design.tools;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+
+import com.xilinx.rapidwright.design.*;
+import com.xilinx.rapidwright.device.PIP;
+import com.xilinx.rapidwright.device.Site;
+import com.xilinx.rapidwright.device.Tile;
+
+public class RelocationTools {
+    public static boolean relocate(Design design,
+                                   String hierarchyPrefix,
+                                   int tileColOffset,
+                                   int tileRowOffset) {
+        Collection<SiteInst> matchingSiteInsts = new ArrayList<>();
+
+        for (SiteInst si : design.getSiteInsts()) {
+            for (Cell c : si.getCells()) {
+                if (c.getName().startsWith(hierarchyPrefix)) {
+                    if (si.isPlaced()) {
+                        matchingSiteInsts.add(si);
+                    }
+                    break;
+                }
+            }
+        }
+
+        return relocate(design, matchingSiteInsts, tileColOffset, tileRowOffset);
+    }
+
+    public static boolean relocate(Design design,
+                                   Collection<SiteInst> siteInsts,
+                                   int tileColOffset,
+                                   int tileRowOffset) {
+
+        HashMap<SiteInst, Site> oldPlacementSite = new HashMap<>();
```

Comment:
Nothing wrong here and there are probably counter examples throughout RapidWright.  But there does exist a `Map` interface in Java that is more generic than `HashMap`.  Perhaps this situation might not need it, but it would allow you to switch to a map that sorts the entries and provides deterministic ordering.

Reply:
I actually reuse `oldPlacementSite` when I move the PIPs to check whether each pin on a net belongs to a matching SiteInst (otherwise it needs to be unrouted).

For this contains operation, a hash map would be more efficient, no? In addition, would I really need deterministic ordering if I'm undo-ing all my placements?

Reply:
Yes, HashMap is probably the right call for the implementation here.  I guess I am referring to the datatype reference which can be made more generic.  Here is an example of what I mean: https://softwareengineering.stackexchange.com/questions/225674/why-define-a-java-object-using-interface-e-g-map-rather-than-implementation